Add support for penalty box substitutions

frank-weinberg opened the pull request:
Also:

- color the (P)LT box field, when a Skater has an unserved penalty
- make the buttons etc. in the popups in the P/LT and SK screens submit immediately
- fix handling of adding past box trips and moving their start/end around (closes #269)
- add possibility to unend box trips

frank-weinberg committed:
Add support for substitutes and annotations

Also mark the box field when a Skater has an unserved penalty

frank-weinberg committed:
Make p/lt popup elemnts immediately submit

also fix NPE when adding a box trip while a jam advance is pending

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -600,4 +623,113 @@ function preparePenaltyEditor() {
+					WS.Set(prefix + 'Annotation', 'Substitute for #' + annotationEditor.data('skaterNumber'));
+					annotationEditor.dialog('close');
+				})).appendTo(row);
+		$('<td>').append($('<button>').text('Self Report').button().click(function() { leaveBox('Self Report');})).appendTo(row);
```
No penalty and Penalty overturned are what the statsbook manual has

(and Left early to be complete)

frank-weinberg replied:
Those are only defined for PBT. But we might as well use them here.

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -600,4 +623,113 @@ function preparePenaltyEditor() {
+
+		var option = $("#AnnotationEditor #subDropdown option[value='"+k.Skater+"']");
+		var inserted = false;
+		if (number != null && option.length == 0) {
```
You can use the appendAlpha functions for this, rather than hand-coding

brian-brazil replied:
This still applies.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/impl/BoxTripImpl.java:
```diff
@@ -231,7 +231,7 @@ public void execute(CommandProperty prop) {
         set(Value.IS_CURRENT, false);
         set(Value.WALLTIME_END, ScoreBoardClock.getInstance().getCurrentWalltime());
-        if (getTeam().hasFieldingAdvancePending()) {
+        if (getTeam().hasFieldingAdvancePending() && getCurrentFielding().getTeamJam().isRunningOrUpcoming()) {
```
There's far more broken than this, and this needs unittests. You should start with the ones I wrote, that covers all the corner cases I found.

frank-weinberg replied:
Instead of vaguely alluding to stuff that's allegedly broken, you should give specific bug reports. And if you want unit tests, provide some that test the intended functionality and not some that test for a feature to be missing.

brian-brazil replied:
The tests I wrote are pretty extensive. It's so broken as is, that 2-3 clicks is enough to show a bug when editing a box trip.

brian-brazil replied:
Put another way, run those unittests by hand and see if they do what you expect. The removal of that behaviour was incidental and was just the way that PR worked out, my goal was to fix the general broken box trip editor.

frank-weinberg replied:
"Your code is a pile of shit" is not exactly motivating me to put any effort into following up. I will not do any work on this unless I get specific bug reports (that are reproducible with the changes from this PR applied). (And if I do change anything, it would be in a separate PR.)

brian-brazil replied:
It's a complex thing to fix so doing it in another PR is fine, though this one should not be marked as fixing #269 as it doesn't.

frank-weinberg replied:
Actually it does. #269 is about adding a skater and a box trip in the same go which was broken as the former was delayed and the latter wasn't. This PR makes adding Skaters immediate, so it addresses this specific scenario. If you have other scenarios, report them properly.

brian-brazil replied:
https://github.com/rollerderby/scoreboard/issues/269#issuecomment-500606708 covers another. And the unittests cover the rest.

frank-weinberg replied:
I don't have the faintest idea what that comment is talking about. But if you insist on keeping that thing open, so be it. I'd rather get this functionality in.

brian-brazil replied:
Create a penalty in the previous jam, advance the start time forward. It'll say Jam 0 at some point.

Considering it's a bugfix, can you add a unittest for this particular change? (Or leave it all for another PR)

frank-weinberg replied:
Ok, I can reproduce that. I have an idea what the cause is, but that's definitely it's own PR.

Everything else should be addressed.

brian-brazil replied:
That's just one form of it, my PR handles all of them so I'd suggest taking a look as I've done most of the leg work already.

brian-brazil commented on html/controls/sk/sk-sheet.js:
```diff
@@ -306,13 +288,29 @@ function prepareTripEditor() {
+							tripEditor.dialog('close');
+						})))
+						.append($('<td width="40%">').append($('<button>').attr('id','insert_before').text('Insert Before').button().click(function() {
+							WS.Set(tripEditor.data('prefix')+'InsertBefore', true);
```
Insert Before is across two lines for me. For consistency with the jam/period editor, Delete should be first.

brian-brazil commented on html/controls/sk/sk-sheet.js:
```diff
@@ -306,13 +288,29 @@ function prepareTripEditor() {
-						.append($('<td width="30%">').append($('<button>').attr('id','submit').text('Submit')))
-						.append($('<td width="40%">').append($('<button>').attr('id','insert_before').text('Insert Before')))
-						.append($('<td width="30%">').append($('<button>').attr('id','remove').text('Remove')))));
+						.append($('<td width="30%">').append($('<button>').attr('id','submit').text('Submit').button().click(function() {
```
It's no longer submitting anything, so Close would make more sense

frank-weinberg replied:
If you removed the score, it still does submit that - the auto submit does not remove the trip (as that would disallow entering a new score to replace the removed one).

brian-brazil replied:
Hmm, maybe require an explicit remove now?

brian-brazil commented on html/controls/sk/sk-sheet.js:
```diff
@@ -306,13 +288,29 @@ function prepareTripEditor() {
+									if (event.which == 13) tripEditor.find('#submit').click(); }).change(function() {
+										WS.Set(tripEditor.data('prefix')+'Score', $(this).val());
+									}))
+								.append($('<button>').attr('id', 'afterSP').text('SP in this or prior trip').button().click(function() {
```
This state isn't propagating correctly on subsequent trips in the UI

frank-weinberg replied:
What does it do and what did you expect?

brian-brazil replied:
If I click on a subsequent trip I expect the SP button to be green, but it isn't.

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -442,17 +455,28 @@ function preparePenaltyEditor() {
+			.append($('<select>').addClass('Period')).append($('<button>').addClass('period_plus').text('+1').button()).appendTo(tr);
+		$('<td width="35%">').append($('<span>').text('Jam: ')).append($('<button>').addClass('jam_minus').text('-1').button())
+			.append($('<select>').addClass('Jam')).append($('<button>').addClass('jam_plus').text('+1').button()).appendTo(tr);
+		$('<td width="15%">').append($('<button>').addClass('set').text('Set Jam').button()).appendTo(tr);
```
This shouldn't be required, changes should be submitted immediately.

frank-weinberg replied:
Since the only way to submit changes to a penalty is via their position in the ordered list and changing the jam number may change that position, changing the jam immediately could result in subsequent changes being submitted for the wrong penalty unless we close the window. But closing the window on every +1/-1 would be incredibly annoying.

brian-brazil replied:
Then the set jam shouldn't show when initially creating a penalty. And it should be jam/period.

brian-brazil replied:
Can the Set Jam button be grouped more with the jam/period? As-is isn't not obvious that they're related. Maybe have Served/Clear on one row? For consistency, Clear should probably be Delete too.

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -442,17 +455,28 @@ function preparePenaltyEditor() {
+				var penaltyNumber = penaltyEditor.data('pnr');
+				var prefix = 'ScoreBoard.Team(' + teamId + ').Skater(' + skaterId + ').Penalty(' + penaltyNumber + ')';
+				WS.Set(prefix + '.ForceServed', active);
+				penaltyEditor.dialog('close');
```
This shouldn't close the dialog.

We also need an explicit close button now

frank-weinberg replied:
When editing a penalty that already exists, it's way more likely that only one attribute is supposed to be changed. So closing on edit is serving the more common workflow.

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -442,17 +455,28 @@ function preparePenaltyEditor() {
-		$('<td>').append($('<input type="checkbox">').attr('id', 'served'))
-			.append($('<span>').text(' has been served')).appendTo(tr2);
+		$('<td>').append($('<button>').text('Served').attr('id', 'served').button().click(function() {
+			var active = $(this).hasClass('checked');
```
This isn't right, it's not showing the WS state

frank-weinberg replied:
It's behaving exactly as it was before.

brian-brazil replied:
No it's not, the button is not showing the state any more.

frank-weinberg replied:
The display was correct. It was setting the value that was broken.

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -338,9 +341,16 @@ function preparePltInputTable(element, teamId, mode, statsbookPeriod, alternateN
+				if (isTrue(WS.state['ScoreBoard.Team('+t+').FieldingAdvancePending'])) {
+					WS.Set('ScoreBoard.Team('+t+').AdvanceFieldings', true);
+				} else if (advanceCell.hasClass('OnTrack')) {
+					openAnnotationEditor(t, id);
```
You can't edit annotations from a jam that has just ended from here, as the advance is showing.

There's also no indication on the LT sheet that there's an annotation.

frank-weinberg replied:
Advance overriding the annotation field was a tradeoff I consciously chose:
- The cases where you need to enter an annotation just as the jam ends are exceedingly rare. And if one arises, you can still enter the annotation in the fielding editor.
- If a user were to enter a substitution before pressing "Advance", the substitution would be entered for the previous interjam. This user error is prevented here.
- There is no other reasonable spot to put either element.

brian-brazil replied:
Okay. Given you're probably on a tablet, typing in "Penalty Overturned" would be tricky. We should probably have buttons for those on the fielding editor too.

brian-brazil commented on html/controls/lt/lt-sheet.js:
```diff
@@ -271,49 +262,69 @@ function prepareFieldingEditor(teamId) {
+		$('<td>').append($('<select>').attr('id', 'skater').append($('<option>').attr('value', '').text('None/Unknown')).change(function() {
+			WS.Set(fieldingEditor.data('prefix')+'Skater', $(this).val());
+		})).appendTo(row);
+		$('<td>').append($('<button>').attr('id', 'notFielded').button().text('No Skater fielded').click(function() {
```
This doesn't show n/a for the upcoming jam until the jam starts

frank-weinberg replied:
It can't be n/a before the jam starts - the team still has the chance to field someone. Thus I've removed the button before Jam start.

brian-brazil commented on html/controls/plt/plt-input.js:
```diff
@@ -600,4 +623,113 @@ function preparePenaltyEditor() {
+					annotationEditor.dialog('close');
+				})).appendTo(row);
+		$('<td>').append($('<button>').text('Self Report').button().click(function() { leaveBox('Self Report');})).appendTo(row);
+		$('<td>').append($('<button>').text('Penalty Revoked').button().click(function() { leaveBox('Penalty Revoked');})).appendTo(row);
```
The revoked doesn't remove the penalty, so it still shows as waiting to serve.

Self report also removes the LT entry, which is incorrect.

frank-weinberg replied:
It's not possible to always correctly remove the penalty: It could be that the trip was for two penalties of which only one was removed. Also if PT and LT are staffed separately, this would allow the LT to meddle with the PT sheet, which I don't think is a good idea.

Self report simply ends the box trip at the point you press it. The only case where a box trip would vanish is if a skater enters and leaves in the same interjam as that means that it never started.

brian-brazil replied:
> It's not possible to always correctly remove the penalty: It could be that the trip was for two penalties of which only one was removed.

In either case it should remove them from the Box.

frank-weinberg replied:
> In either case it should remove them from the Box.

It does.

brian-brazil replied:
Hmm, can't reproduce now.

frank-weinberg committed:
review comments

frank-weinberg committed:
Add penalty counts to LT screen

Also make highlighting the existence of unserved penalties actually work
on LT and more reliable on PT

brian-brazil commented:
Something else I noticed is that when a jam is ongoing and you click Box, the lineup symbols flicker to S before becoming -.

frank-weinberg committed:
more review comments

brian-brazil commented on html/controls/lt/lt-sheet.js:
```diff
@@ -271,61 +263,81 @@ function prepareFieldingEditor(teamId) {
+		row = $('<tr>').addClass('Skater').appendTo(table);
+		$('<td>').attr('colspan', '3')
+			.append($('<input type="text">').attr('size', '35').attr('id', 'annotation'))
+			.append($('<button>').attr('id', 'setAnnotation').text('Set').button().click(function() {
```
Do we really need this Set button? I'd expect this to be submitted automatically via an oninput handler.

brian-brazil reviewed (commented):
The S is still flickering when you press Box during a jam.

I also noticed that if you're between periods and the intermission clock has expired, you can't create new penalties - but this appears to be pre-existing.

frank-weinberg committed:
Make annotation fields submit on typing

frank-weinberg commented:
I've removed the set buttons. But the flickering S is an artefact of the fact that the EventListeners are run before `valueChanged`. Changing that order would be guaranteed to break stuff elsewhere in the software.

brian-brazil commented:
That sounds like the lack of a batch somewhere.

brian-brazil commented:
...and that looks like something I've already on my todo list.

brian-brazil commented:
There's still no indication on the LT sheet that a given fielding has an annotation.

When a skater is not in the box, and you click the ... the Penalty Overturned etc. aren't present.

frank-weinberg committed:
add annotation indiciation

frank-weinberg commented:
> That sounds like the lack of a batch somewhere.

This has nothing to do with batches.

> There's still no indication on the LT sheet that a given fielding has an annotation.

Now there is.

> When a skater is not in the box, and you click the ... the Penalty Overturned etc. aren't present.

Of course not. There's no penalty that could be overturned.

brian-brazil commented:
> Of course not. There's no penalty that could be overturned.

The skater may have already left the box, but you still want to add the annotation. I'd expect the interface to be the same whether they've left the box or not.

frank-weinberg commented:
The LT input interface is used to maintain the current game state. If you want to edit past events, you use the sheet part.

brian-brazil commented:
The current jam is current state though, and annotations are tied to fieldings not penalties.

frank-weinberg commented:
The semantics of the buttons is "End the current box trip and add the reason as an annotation." That is tied to the box trip.

brian-brazil commented:
Having one place where you can always press a button to add to the text, and another place where you sometimes can and sometimes can't is likely to cause confusion. For as long as these are tied to fieldings (which I think is the way to handle it) whether they're in the box or not shouldn't matter.

frank-weinberg committed:
Add possibility to unend box trips

in passing fix handling of moving start/end of box trips and add unit
tests so the reviewer is satisfied. Closes #269

frank-weinberg commented:
Ok, fine. So now both sets of buttons are available if and only if there is a box trip associated with the fielding.

And since they were cumbersome to do in parallel, I also included the box trip fixes in here and even copied your unit tests (though I don't think checking the paperwork symbols is the best way to do them).

brian-brazil commented on tests/com/carolinarollergirls/scoreboard/core/impl/FieldingImplTests.java:
```diff
@@ -85,4 +86,52 @@ public void testSortBoxTripSymbols() {
+       assertEquals(1, f1.getAll(Fielding.Child.BOX_TRIP).size());
+       assertEquals(1, f2.getAll(Fielding.Child.BOX_TRIP).size());
+
+        // And remed again.
```
removed

brian-brazil reviewed (commented):
I was away the past week, so am only getting to this now.

I only see one bug from testing now, the red dot indicating an annotation doesn't show for the upcoming jam.

brian-brazil commented:
> though I don't think checking the paperwork symbols is the best way to do them

Feel free to do something better, though the tests for the symbols themselves were also pretty light.

frank-weinberg committed:
Add annotation indicator for upcoming jam

frank-weinberg commented:
Annotation indicator for the upcoming jam has been added.
